client/state/happychat/middleware.js

```diff
@@ -7,21 +7,86 @@ import throttle from 'lodash/throttle';
 /**
  * Internal dependencies
  */
+import wpcom from 'lib/wp';
 import {
+	HAPPYCHAT_CONNECTING,
 	HAPPYCHAT_SEND_BROWSER_INFO,
 	HAPPYCHAT_SEND_MESSAGE,
 	HAPPYCHAT_SET_MESSAGE,
 	HAPPYCHAT_TRANSCRIPT_REQUEST,
 } from 'state/action-types';
-import { receiveChatTranscript } from './actions';
-import { getHappychatTranscriptTimestamp } from './selectors';
+import {
+	setChatConnected,
+	receiveChatEvent,
+	receiveChatTranscript,
+	requestChatTranscript,
+	setHappychatAvailable,
+	setHappychatChatStatus,
+} from './actions';
+import {
+	getHappychatConnectionStatus,
+	getHappychatTranscriptTimestamp,
+} from './selectors';
+import {
+	getCurrentUser,
+	getCurrentUserLocale,
+} from 'state/current-user/selectors';
 
 const debug = require( 'debug' )( 'calypso:happychat:actions' );
 
 const sendTyping = throttle( ( connection, message ) => {
 	connection.typing( message );
 }, 1000, { leading: true, trailing: false } );
 
+// Promise based interface for wpcom.request
+const request = ( ... args ) => new Promise( ( resolve, reject ) => {
+	wpcom.request( ... args, ( error, response ) => {
```

Comment:
Absolutely. Since this is just a copy-paste to move already-working code from one place to another, do you mind if I do that in another PR? I'll add it to the list on the open issue about the middleware refactor.